legion/engine/physics/halfedgeedge.cpp

```diff
@@ -0,0 +1,243 @@
+        //             [face of 'this']           [merge face]                  //
+        //                                                                      //
+        //----------------------------------------------------------------------//
+
+        auto handleDoubleAdjacentMergeResult = [this,releaseFaceToVert](HalfEdgeEdge* prevFromCurrent, HalfEdgeEdge* prevFromCurrentConnection)
+        {
+            HalfEdgeFace* invariantMergeFace = prevFromCurrent->pairingEdge->face;
+            releaseFaceToVert(invariantMergeFace);
+
+            HalfEdgeEdge* prevFromPFC = prevFromCurrent->prevEdge;
+            HalfEdgeEdge* newNextFromPFC = prevFromCurrent->pairingEdge->nextEdge;
+            prevFromPFC->setNext(newNextFromPFC); 
+
+            HalfEdgeEdge* nextFromFCC = prevFromCurrentConnection->nextEdge;
+            HalfEdgeEdge* newPrevFromFCC = prevFromCurrentConnection->pairingEdge->prevEdge;
+            nextFromFCC->setPrev(newPrevFromFCC); 
+
+            delete prevFromCurrent->pairingEdge;
+            delete prevFromCurrent;
+
+            delete prevFromCurrentConnection->pairingEdge;
+            delete prevFromCurrentConnection;
+
+            face->startEdge = prevFromPFC;
+
+            std::vector<math::vec3> vertices;
+            auto collectVertices = [&vertices](HalfEdgeEdge* currentEdge) { vertices.push_back(currentEdge->edgePosition); };
+            face->forEachEdge(collectVertices);
+
+            float tempDis;
+            PhysicsStatics::calculateNewellPlane(vertices, face->normal, tempDis);
+
+            invariantMergeFace->startEdge = nullptr;
+            delete invariantMergeFace;
+        };
+
+        HalfEdgeFace* currentFace = face;
+        auto setEdgeFace = [&currentFace](HalfEdgeEdge* edge) {edge->face = currentFace; };
+
+        if (prevFromCurrent->pairingEdge->face == prevFromCurrentConnection->pairingEdge->face)
+        {
+            face->forEachEdge(setEdgeFace);
+
+            math::vec3 norm;
+            if (PhysicsStatics::isNewellFacesCoplanar(face, prevFromCurrent->pairingEdge->face, prevFromCurrent, scalingEpsilon, norm,2))
+            {
+                handleDoubleAdjacentMergeResult(prevFromCurrent, prevFromCurrentConnection);
+            }
+        }
+
+        if (nextFromCurrent->pairingEdge->face == nextFromCurrentConnection->pairingEdge->face)
+        {
+            face->forEachEdge(setEdgeFace);
+
+            math::vec3 norm;
+            if (PhysicsStatics::isNewellFacesCoplanar(face, nextFromCurrent->pairingEdge->face, nextFromCurrentConnection, scalingEpsilon, norm,2))
+            {
+                handleDoubleAdjacentMergeResult(nextFromCurrentConnection, nextFromCurrent);
+            }
+        }
+
+        //[3] re-Initialize to account for new edges
+        face->initializeFace();
+
+        //[4] delete pairing and 'this' edge. These edges are no longer part of this face
+        delete pairingEdge;
+        delete this;
```

Comment:
my justification here is that my use of delete this is based on the fact I conform with the requirements mention in ( https://isocpp.org/wiki/faq/freestore-mgmt#delete-this ), But this will be fixed when we switch to smart pointers in a later pr

Reply:
yeah well the problem with that is that with an open codebase like this we can't guarantee that no one might want to try to use anything in a different use case at some point. and generally it's also bad code design to have something self deleting.
It would imply the object is self owning, but it did not create itself. so this does not comply to RAII and ownership principles